File: Project_A/srcFilesForStudents_ProjectA/linecount.c
// linecount.c - Count lines in a sourcefile

#include <stdio.h>

int main(int argc, char* args[]) {
    FILE* srcFile = NULL;
    char* srcFilename = "<srcFilename>";

    // Check the command-line.
    if (argc != 2) {
        printf("Usage: linecount <src>\n");
        return 0;
    }

    // Check if arguments are valid, if the srcFile exists, and if can create the dstFile.
    if (args[1] != NULL) { // Check <src>
        srcFilename = args[1];
//      printf("linecount: srcFilename = '%s'\n", srcFilename);
        srcFile = fopen(srcFilename, "r");
        if (srcFile == NULL) {
            printf("linecount: Cannot open srcFile '%s'\n", srcFilename);
            return 0;
        }
    } else {
        printf("linecount: [OK] srcFilename = '%s'\n", srcFilename);
    }

    // Execute the line count.
    int c;
    int nLines = 0;

    while ((c = fgetc(srcFile)) != EOF) {
        if (c == '\n') ++nLines;
    }

    // Close and flush the file.
    fclose(srcFile);

    // Print the result.
    printf("%d lines\n", nLines);
    return 1;
}

File: Project_A/src/LineCount.java
import java.io.*;

public class LineCount {

    private static int EOF = -1;
    private static File srcFile = null;
    private static String srcFilename = "<srcFilename>";

    private static void helper() {

        System.out.println("\n");
        System.out.println("CharCount:\t" + "Usage: java CharCount.java <src>");
        System.out.println("Copy:\t\t" + "Usage: java Copy.java <src> <dst>");
        System.out.println("LineCount:\t" + "Usage: java LineCount.java <src>");
        System.out.println("WordCount:\t" + "Usage: java WordCount.java <src>");
        System.out.println("WC:\t\t" + "Usage: java WC.java <src> { <src> }");

        System.out.println("\n");

    }

    public static void main(String[] args) throws IOException {

        if (args.length != 1) {
            System.out.println("Usage: linecount <src>\n");
            return;
        }

        boolean isHelper = args[0].equals("-?") || args[0].equals("-h") || args[0].equals("-help");

        if (isHelper) {
            helper();
            return;
        }

        if (args[0] != null) { // Check <src>
            srcFilename = args[0];
            System.out.println("linecount: srcFilename = " + srcFilename + "");
            srcFile = new File(srcFilename);
            if (!srcFile.canRead()) {
                System.out.println("linecount: Cannot open srcFile '" + srcFilename + "'");
                return;
            }
        } else {
            System.out.println("linecount: [OK] srcFilename = '" + srcFilename + "'");
        }

        final FileInputStream srcStream = new FileInputStream(srcFile);
        // Execute the character count.
        int c;
        int nLines = 1;

        while ((c = srcStream.read()) != EOF) {
            if (c == '\n')
                nLines++;
        }

        srcStream.close();

        System.out.printf("%d line(s)\n", nLines);

    }

}
